add notification module for Microsoft Teams

diadzine opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
teams

##### ANSIBLE VERSION
```
ansible 2.3.0 (devel 0224760725) last updated 2017/03/02 23:56:33 (GMT +200)
  config file = 
  configured module search path = Default w/o overrides
```

##### SUMMARY
New module to send notification to Microsoft Teams.

diadzine committed:
add notification module for Microsoft Teams

mattclay commented:
The test `ansible-test sanity --test validate-modules` failed with the following errors:

```            
lib/ansible/modules/notification/teams.py:0:0: E312 No RETURN documentation provided
lib/ansible/modules/notification/teams.py:0:0: E208 module_utils imports should import specific components, not "*". line 139
lib/ansible/modules/notification/teams.py:0:0: E208 module_utils imports should import specific components, not "*". line 140
lib/ansible/modules/notification/teams.py:0:0: E107 Imports should be directly below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA. line 139
lib/ansible/modules/notification/teams.py:0:0: E107 Imports should be directly below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA. line 140
```

diadzine committed:
Additional changes to pass CI

Add RETURN documentation.
Import specific components from module_utils instead of '*'.
Move imports directly below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA.

diadzine committed:
Fix typo in ANSIBLE_METADATA.version

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,151 @@
+      - You can also control a color bar by setting color to a color hex code.
+        This bar appears to the left of the cards content, and is primarily
+        used to indicate status to the user.
+    required: false
```
Parameters that are not required don't need `required: false`.

dagwieers replied:
Parameters that are not required don't need `required: false`.

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,151 @@
+        This bar appears to the left of the cards content, and is primarily
+        used to indicate status to the user.
+    required: false
+    default: None
```
If there's no default, `None' is implicit. (It means parameter undefined)
So this is not needed.

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,151 @@
+        argument_spec = dict(
+            url = dict(type='str', required=True, no_log=True),
+            text = dict(type='str', required=True),
+            title = dict(type='str', required=False, default=None),
```
Same here. required=False and default=None are obsolete. Only specify them if they are set.

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,151 @@
+            url = dict(type='str', required=True, no_log=True),
+            text = dict(type='str', required=True),
+            title = dict(type='str', required=False, default=None),
+            color = dict(type='str', required=False, default=None)
```
Same here. required=False and default=None are obsolete. Only specify them if they are set.

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,151 @@
+            title = dict(type='str', required=False, default=None),
+            color = dict(type='str', required=False, default=None)
+        )
+    )
```
Please implement check-mode. In this case check-mode could either not send a notification, or better, sends a "fake" notification (so authentication is attempted). It depends on the API if this is feasible. It's always better that check-mode is performing more logic without actually doing something.

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,151 @@
+  title:
+    description:
+      - A message can also contain a title.
+    required: false
```
Parameters that are not required don't need `required: false`.

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,151 @@
+    description:
+      - A message can also contain a title.
+    required: false
+    default: None
```
If there's no default, `None' is implicit. (It means parameter undefined)
So this is not needed.

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,151 @@
+    payload = build_payload_for_teams(module, text, title, color)
+    do_notify_teams(module, url, payload)
+
+    module.exit_json(msg="OK")
```
What's the purpose of sending `msg="OK"` if the success of the task indicates the same thing ?

diadzine committed:
Additional changes after code review

Remove required = false and default = None.
Remove required=false, defautl=None.
Remove useless return variable msg="OK".
Implement check_mode.

bjolivot commented:
Thank you for your module :+1: 
shipit

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,147 @@
+        module.exit_json(changed=True)
+    else:
+        module.fail_json(
+            msg=" failed to send %s to Teams incoming webhook: %s" % (
```
Not sure why the error starts with a whitespace here.

gundalow commented:
Hi, there is an alternative version https://github.com/ansible/ansible/pull/23277

Would be good if between you both you could decide which should go forward.

marc-sensenich commented:
@diadzine I am happy to share ownership and maintenance moving forward on this module's functionality and am open to merging whichever PR accomplishes this best.

northalpha commented:
@marc-sensenich could you please include your message cards support? that would be very nice! thanks!

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,147 @@
+      U(https://teams.microsoft.com/start) through Office 365 Connectors via
+      the Incoming WebHook integration.
+version_added: "2.3"
+author: "Aymeric Bringard (@diadzine)"
```
Make this a proper list:
```yaml
author:
- Aymeric Bringard (@diadzine)
```

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,147 @@
+    title = module.params['title']
+    color = module.params['color']
+
+    if module.check_mode:
```
Move the check_mode test into the `do_notify_teams` function, and see if you can do something relevant instead e.g. test the connectivity so you do as much as possible in check-mode.

dagwieers commented:
If you want to speed up getting this merged, put it on the agenda of the core meeting and ask for a review there.

+label new_module

marc-sensenich commented:
@dagwieers this isn't my pull request, I ended up opening #23277 before I saw this one which implemented all the features of the 365 connector cards at the time. I'm willing to reopen a PR with my module implementation and put it on the agenda since this PR has gone stale. Where would I submit an agenda item?

dagwieers commented:
@marc-sensenich Did you talk to @diadzine ? It's best to come to an agreement.

marc-sensenich commented:
I reached out in April in this PR as a comment after I opened mine and had no response.

dagwieers commented:
@marc-sensenich I feel your pain. So maybe open your PR again, and work on that one. Then add a comment on the agenda about this situation and we will discuss in the meeting when you join.

The meeting agenda is here: https://github.com/ansible/community/issues/162

jhawkesworth commented:
Thanks for this module.  

I just tried this out and it works for me so it would be great to get this (or a more enhanced version) included in ansible.

jhawkesworth commented:
@diadzine - Are you still interested in getting this module included in ansible?  There's a little bit of time to respond to the feedback above and fix this up before ansible 2.4 deadline.  If you don't have time for this please can you respond as there are others willing to make and maintain a module for Teams.  Many thanks, Jon

s-hertel commented:
@diadzine @dagwieers @jhawkesworth 
Since there has been no activity updating this since March and it would be a good thing to make it into 2.4 and @marc-sensenich's PR in addition has more functionality, I would like to merge #27636 (after pep8 is fixed). Are there any objections, or do people still want a chance to review it?

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/python
```
Remove this shebang (CI will fail)

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,147 @@
+
+# (c) 2017, Aymeric Bringard <diadzine@gmail.com>
+#
+# This file is part of Ansible
```
We now have a one-line GPL stub for modules. See other modules in devel branch.

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,147 @@
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'status': ['preview'],
```
This is the old ANSIBLE_METADATA stub. CI will fail.

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,147 @@
+      - The contents of text will be displayed in a card format, along with a
+        sender image and display name. This image and display name are set
+        when the webhook is configured.
+        The simplest message just contains a text variable.
```
Make this a separate item in the list.

dagwieers commented on lib/ansible/modules/notification/teams.py:
```diff
@@ -0,0 +1,147 @@
+        sender image and display name. This image and display name are set
+        when the webhook is configured.
+        The simplest message just contains a text variable.
+        To add links to text, developers **must** include Markdown syntax for
```
Make this a separate item in the list.

dagwieers commented:
@s-hertel I had some more unanswered comments, and added a few more (related to future CI failures). But I cannot test the module itself. For me this is fine to be merged if it works for others.

s-hertel commented:
@dagwieers in regard to #27636? I think #27636 should be merged since marc-sensenich is active and happy to make fixes.

dagwieers commented:
@s-hertel Yes, sorry somehow I misread your comment. This one can be closed IMO.

diadzine commented:
Sorry guys, I was busy those last months with no time on this PR.
As @s-hertel suggested #27636 has more functionality and @marc-sensenich is more active than me ;)
I close this PR.
